### PR TITLE
Lgb modelers

### PR DESCRIPTION
**Pull request by ajain-ida**
Created first draft of lgb modelers. No validation set usage or hyper optimization

### CODE REVIEW

File: fifeforspark/lgb_modelers.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            time_horizon: int,
+            params: Union[None, dict] = None,
+            subset: Union[None, pyspark.sql.column.Column] = None
+    ) -> pyspark.ml.pipeline.PipelineModel:
+        """Train a LightGBM model for a single lead length."""
+        if params is None:
+            params = {
+                time_horizon: {
+                    "objective": self.objective,
+                    "num_iterations": self.config.get("MAX_EPOCHS", 256),
```
Comment (EdJishWang) [inline fifeforspark/lgb_modelers.py:93]: Max_epochs isn't in the argument parser yet, do you mind adding it?

File: fifeforspark/lgb_modelers.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            self.n_intervals = self.set_n_intervals()
+        self.model = self.train(
+            params=params
+        )
+
+    def train(
+            self,
+            params: Union[None, dict] = None,
+            subset: Union[None, pyspark.sql.column.Column] = None,
+            validation_early_stopping: bool = True,
```
Comment (EdJishWang) [inline fifeforspark/lgb_modelers.py:67]: 'Validation_early_stopping' argument not being used

File: fifeforspark/lgb_modelers.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                if cumulative:
+                    predictions = predictions.withColumn(f'probability_{i + 1}',
+                                                         predictions[f'probability_{i + 1}'] *
+                                                         predictions[f'probability_{i}'])
+        return predictions
+
+    def transform_features(self) -> pyspark.sql.DataFrame:
+        """Transform features to suit model training."""
+        data = self.data
+        date_cols = [x for x, y in data.dtypes if y in ['DateType', 'Timestamp']]
```
Comment (EdJishWang) [inline fifeforspark/lgb_modelers.py:156]: Date and Timestamp types resulting from data.dtypes should be ‘date’ and ‘timestamp’ (all lowercase for both). So: [‘date', 'timestamp']

File: fifeforspark/lgb_modelers.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        for col in date_cols:
+            data = data.withColumn(col,
+                                   10000*date_format(data[col], "y") +
+                                   100*date_format(data[col], "m") +
```
Comment (EdJishWang) [inline fifeforspark/lgb_modelers.py:160]: Lowercase ‘m’ stands for minutes in Pyspark datetimes. To get the months I think we use ‘M’.

File: fifeforspark/lgb_modelers.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        predictions = first_model.transform(predict_data).selectExpr('probability as probability_1')
+        predictions = predictions.withColumn('probability_1', firstelement(predictions['probability_1']))
+        for i, lead_specific_model in enumerate(self.model):
+            if i != 0:
+                predictions = predictions.withColumn(f'probability_{i+1}',
+                                                     firstelement(predictions[f'probability_{i+1}']))
+                if cumulative:
+                    predictions = predictions.withColumn(f'probability_{i + 1}',
+                                                         predictions[f'probability_{i + 1}'] *
+                                                         predictions[f'probability_{i}'])
```
Comment (EdJishWang) [inline fifeforspark/lgb_modelers.py:150]: I might be wrong, but it doesn’t seem like you’re calling .transform() on anything past the first model (no transforms on ‘lead_specific_model’ within the enumerate loop).

**Review by EdJishWang — CHANGES_REQUESTED** [review]
Really nice job! Just a few potential changes.

File: fifeforspark/lgb_modelers.py (hunk 3 above)
Comment (ajain-ida) [reply fifeforspark/lgb_modelers.py:156]: ahhh I forgot to change this one, good catch

**Comment by ajain-ida** [comment]
Merged in lgbmodelers

**Pull request merged**